### PR TITLE
ARROW-10366: [Rust][DataFusion] Do not buffer intermediate results in merge or HashAggregate

### PR DESCRIPTION
**Pull request by alamb**
This PR is based on https://github.com/apache/arrow/pull/8503 from @jorgecarleitao 

This makes `merge` send batches to a receiver stream as they arrive and, thereby removing the need to wait for each thread to finish collecting all its batches. It improves reported performance in my laptop on the aggregate_query_sql microbenchmark by 5-15%. 

It also includes a change to the aggregate code that allows the aggregates to produce output even if there are no results on the first call to `poll_next`. The change in GroupBy is needed because when we initially tried to run Merge in parallel, we discovered no output was created if the inputs had not completed producing all their values *before* execution began. There is much more discussion on the topic here: https://github.com/apache/arrow/pull/8503#discussion_r510061337. I will add additional comments on the changes inline below. 

# Performance

I measured Aggregate Performance with the following command
```
git checkout master && \
    cargo bench --bench aggregate_query_sql && \
    git checkout alamb/ARROW-10366-collect-in-parallel && \
    cargo bench --bench aggregate_query_sql
```

<details>
  <summary>Click for performance details</summary>

aggregate_query_no_group_by 15 12
                        time:   [715.38 us 717.64 us 720.07 us]
                        change: [-5.6870% -4.7352% -3.6834%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 8 outliers among 100 measurements (8.00%)
  2 (2.00%) high mild
  6 (6.00%) high severe

Benchmarking aggregate_query_group_by 15 12: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 13.7s or reduce sample count to 40.
aggregate_query_group_by 15 12
                        time:   [2.4874 ms 2.5029 ms 2.5173 ms]
                        change: [-13.513% -12.381% -11.184%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  3 (3.00%) high mild
  2 (2.00%) high severe

Benchmarking aggregate_query_group_by_with_filter 15 12: Warming up for 3.0000 s
Warning: Unable to complete 100 samples in 5.0s. You may wish to increase target time to 10.2s or reduce sample count to 40.
aggregate_query_group_by_with_filter 15 12
                        time:   [1.9724 ms 1.9856 ms 1.9982 ms]
                        change: [-14.759% -10.342% -6.5522%] (p = 0.00 < 0.05)
                        Performance has improved.
Found 5 outliers among 100 measurements (5.00%)
  2 (2.00%) high mild
  3 (3.00%) high severe

</details>

### CODE REVIEW

File: rust/datafusion/Cargo.toml [hunk 1]
```diff
@@ -56,6 +56,7 @@ num_cpus = "1.13.0"
 chrono = "0.4"
 async-trait = "0.1.41"
 futures = "0.3"
+pin-project-lite= "^0.1.1"
```
Comment (alamb) [inline rust/datafusion/Cargo.toml:59]: Note that this does not add any new actual dependencies (pin_project_lite is already required indirectly via tokio)

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 2]
```diff
@@ -214,13 +217,13 @@ Example: average
...
 * Finally, `get_value` returns an array with one entry computed from the state
 */
-struct GroupedHashAggregateStream {
-    mode: AggregateMode,
-    schema: SchemaRef,
-    group_expr: Vec<Arc<dyn PhysicalExpr>>,
-    aggr_expr: Vec<Arc<dyn AggregateExpr>>,
-    input: SendableRecordBatchStream,
-    finished: bool,
+pin_project! {
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:220]: I found the whole notion of having to Pin things to call `poll_next` on streams to be unnaturally confusing. While I basically understand what `pin_project!` does, I will admit I do not understand why Streams need to be `Pin`. I am just cargo culting it there

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 3]
```diff
@@ -315,6 +318,41 @@ fn group_aggregate_batch(
     Ok(accumulators)
 }
 
+async fn compute_grouped_hash_aggregate(
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:321]: The main change in this file is to pull the computation of the output record batch into a future (aka an async function). I think this formulation is still a natural way to express the logic (you can see each input batch being fed to `group_aggregate_batch`)

File: rust/datafusion/src/physical_plan/hash_aggregate.rs [hunk 4]
```diff
@@ -324,12 +362,24 @@ impl GroupedHashAggregateStream {
         aggr_expr: Vec<Arc<dyn AggregateExpr>>,
         input: SendableRecordBatchStream,
     ) -> Self {
+        let (tx, rx) = futures::channel::oneshot::channel();
+
+        let schema_clone = schema.clone();
+        tokio::spawn(async move {
```
Comment (alamb) [inline rust/datafusion/src/physical_plan/hash_aggregate.rs:368]: this spawns a separate task to do the actual aggregation (and drive the inputs, as needed, by calling `input.next().await`). Driving the input was not happening when using `future.into_stream` in the actual call to `poll_next` https://github.com/apache/arrow/pull/8553/files#diff-4dbe750a2836f6c723bc36ae39f2faf15e2706303f5dbb042ae00f1f36658c62L399

**Comment by andygrove** [comment]
I will try and run some TPC benchmarks tomorrow. I'm excited to see how this looks. Great work @jorgecarleitao and @alamb

File: rust/datafusion/src/physical_plan/hash_aggregate.rs (hunk 2 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:220]: The crucial observation is that a stream is not an iterator of futures - it is the asynchronous version of an iterator. 

A stream needs to be pinned for the same reason a `Future` [needs to be pinned](https://rust-lang.github.io/async-book/04_pinning/01_chapter.html#why-pinning): the scheduler can stop the work of a `stream` and point to its state via a reference, and the only way for the reference to not become invalid is to forbid the memory from being physically moved (via the `Pin` marker).

File: rust/datafusion/Cargo.toml (hunk 1 above)
Comment (jorgecarleitao) [reply rust/datafusion/Cargo.toml:59]: You did well. This IMO is the correct way to approach this. All relevant async projects use this lib, and even std uses a variation of it AFAIK.

File: rust/datafusion/src/physical_plan/hash_aggregate.rs (hunk 3 above)
Comment (jorgecarleitao) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:321]: I also felt the need for this at some of my tries to simplify the hashAgg. code. 👍

**Review by jorgecarleitao — COMMENTED** [review]
I went through this. Beautiful work, @alamb !

My only comment is that we are performing a change in the git submodule `testing` which I do not think it was intended.

All the rest is really cool stuff and makes a lot of sense.

I also look forward for the large benchmarks!

**Comment by alamb** [comment]
> I went through this. Beautiful work, @alamb !

Thanks @jorgecarleitao 

> My only comment is that we are performing a change in the git submodule testing which I do not think it was intended.

You are correct that that was not intended -- I will fix that

File: rust/datafusion/src/physical_plan/hash_aggregate.rs (hunk 2 above)
Comment (alamb) [reply rust/datafusion/src/physical_plan/hash_aggregate.rs:220]: @jorgecarleitao  -- that is a beautifully eloquent way of explaining the need for Pin. Thank you for that.

**Comment by andygrove** [comment]
Here are the TPC benchmark results (sf=100) based on the following command.

```bash
cargo run --release --bin tpch -- --iterations 3 --path /mnt/tpch/parquet-100GB --format parquet --query 1 --batch-size 4096 --concurrency 24
```

### 2.0.0 release

```
Query 1 iteration 0 took 13779 ms
Query 1 iteration 1 took 13665 ms
Query 1 iteration 2 took 14008 ms
```

### master branch

```
Query 1 iteration 0 took 20304 ms
Query 1 iteration 1 took 20228 ms
Query 1 iteration 2 took 20737 ms
```

### This PR

```
Query 1 iteration 0 took 20528 ms
Query 1 iteration 1 took 20688 ms
Query 1 iteration 2 took 20432 ms
```

I'm not sure why there is such a regression since 2.0.0 and will see if I can track down where this started.

**Comment by andygrove** [comment]
The performance regression is related to removing use of the specialization trait. I filed https://issues.apache.org/jira/browse/ARROW-10453

**Comment by alamb** [comment]
@andygrove 

Do you have the instructions for generating (or would be willing to provide me a copy to) the 100GB parquet dataset you are using in the following benchmark? I would love to try and run it locally and do some profiling to understand more of where time is being spent.

```
cargo run --release --bin tpch -- --iterations 3 --path /mnt/tpch/parquet-100GB --format parquet --query 1 --batch-size 4096 --concurrency 24
```

Thanks!
Andrew

**Comment by andygrove** [comment]
@alamb I've been using Apache Spark to do the conversion and re-partitioning, using code in the Ballista project that I've been running manually from my IDE. I wouldn't feel good about checking Scala code into the Arrow Rust project so maybe the short term solution is for me to put a Dockerfile in Ballista for creating a Docker image that can do the conversion. Long term we should write a Rust utility to read CSV and write to Parquet. I'm not sure if the Parquet writer is far enough along yet to support that?

**Comment by andygrove** [comment]
@alamb Could you try this out. You'll need to map volumes in appropriately to read/write from host directories.

```bash
docker run -it ballistacompute/spark-benchmarks:0.4.0-SNAPSHOT 
  -h, --help   Show help message

Subcommand: convert-tpch
  -i, --input  <arg>
      --input-format  <arg>
  -o, --output  <arg>
      --output-format  <arg>
  -p, --partitions  <arg>
  -h, --help                   Show help message
Subcommand: tpch
      --input-format  <arg>
  -i, --input-path  <arg>
  -q, --query  <arg>
  -h, --help                  Show help message
```

**Comment by alamb** [comment]
> @alamb Could you try this out. You'll need to map volumes in appropriately to read/write from host directories.

Checking it out...

**Comment by alamb** [comment]
@andygrove  -- it seems to work well:

```
$ docker run  -it -v /Users/alamb/Software/tpch_data/SF1:/opt/spark/work-dir ballistacompute/spark-benchmarks:0.4.0-SNAPSHOT convert-tpch  --input lineitem.tbl --input-format tbl --output lineitem.parquet --output-format parquet --partitions 16
```

Which then produced output like this:

```
$ ls -l /Users/alamb/Software/tpch_data/SF1/lineitem.parquet/
total 428032
-rw-r--r--  1 alamb  staff     0B Nov  2 06:27 _SUCCESS
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00000-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00001-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00002-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00003-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00004-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00005-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00006-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00007-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00008-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00009-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00010-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00011-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00012-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00013-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00014-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
-rw-r--r--  1 alamb  staff    13M Nov  2 06:27 part-00015-47e26762-3d8a-4652-ba5e-2e2a620e399e-c000.snappy.parquet
```

I plan to test out using these files with datafusion later today if I have time, or tomorrow if not. 

Thanks again,
Andrew

**Comment by alamb** [comment]
@jorgecarleitao  and @andygrove  -- I did some profiling of the tpch benchmark (SF10) locally on my laptop this afternoon. I was not able to reproduce a slowdown on this branch

Here is my data from on my laptop (a Macbook Pro):

Test command was: `cargo run --release --bin tpch -- --iterations 3 --path /Users/alamb/Software/tpch_data/SF10-parquet-64 --format parquet --query 1 --batch-size 4096`

When run against  @ d4121d8a17d9e53ad4421960e357dd2f89771603

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 2, batch_size: 4096, path: "/Users/alamb/Software/tpch_data/SF10-parquet-64", file_format: "parquet", mem_table: false }
Query 1 iteration 0 took 8913 ms
Query 1 iteration 1 took 10533 ms
Query 1 iteration 2 took 10633 ms
```

When run against this branch @ 802868617777022bd0945f98f987993ca7267eb6 (after rebasing against d4121d8a17d9e53ad4421960e357dd2f89771603):

```
Running benchmarks with the following options: TpchOpt { query: 1, debug: false, iterations: 3, concurrency: 2, batch_size: 4096, path: "/Users/alamb/Software/tpch_data/SF10-parquet-64", file_format: "parquet", mem_table: false }
Query 1 iteration 0 took 9168 ms
Query 1 iteration 1 took 10097 ms
Query 1 iteration 2 took 10199 ms
```

So in other words I didn't measure a noticeable change on my machine

Qualatatively both this branch and master While kept my CPUs ~ 100% busy and the memory used was fairly consistent

I then spent some time poking around with a profile from XCode Instruments to see where the time was going. Unsurprisingly it is going to memory management (this is the story of my professional life):

Screen Shot 2020-11-05 at 5 39 03 PM

I am going to poke some more time seeing if I can figure out any easy to recycle buffers or find a smoking gun

**Comment by andygrove** [comment]
Thanks @alamb for posting the detailed info. I hope I haven't caused confusion with my earlier comments and performance data. I was not seeing a slowdown on this PR compared to master. I was seeing a slowdown in master compared to the 2.0.0 release (which I tracked down to the commit that removed specialization).

**Comment by alamb** [comment]
Ah -- got it. 👍

**Review by jorgecarleitao — APPROVED** [review]
Did we conclude why this does not improve performance on the large benchmark?

For me, the code is good as is, but let me know if there is some investigation in progress on this.

**Comment by alamb** [comment]
@jorgecarleitao -- when I ran the TPCH benchmark Q1 locally on my machine, I found it kept all my cores busy and the memory profile was low. Thus the improvements offered by this PR of running more things in parallel and avoiding buffering don't address the (current) bottleneck in the TPCH query

I theorize (without presenting evidence yet) that this is due to how the query is running and that:
1. Runtime is  dominated by filtering (not aggregation)
2. The intermediate results are small (the output of the first stage of grouping is 4 rows)

I have a more detailed performance analysis half written up but it is not coherently enough to share yet . I am hoping for some time later this week to write it up

I (selfishly) think this is an improvement, but I don't have any performance numbers to back that up

**Comment by alamb** [comment]
Rebased

**Review by andygrove — APPROVED** [review]
LGTM.

**Comment by rdettai** [comment]
I had some code that was crashing because of the behavior aggregate had when the wrapped exec first returned `Pending` when being polled. It know works perfectly with this PR! Thanks Andrew! Once this is merged I have some tests ready in #8658 that validate this!

**Comment by alamb** [comment]
I plan to merge this tomorrow unless i hear otherwise. @jorgecarleitao  / @andygrove  let me know if you have any concerns

**Comment by alamb** [comment]
Rebased

**Comment by alamb** [comment]
The [travis CI](https://github.com/apache/arrow/pull/8553/checks?check_run_id=1399712911) checks (which aren't for rust specific things) seem to be stuck / not working. I don't think they are related to this PR (e.g. the same thing happened on [This unrelated PR](https://github.com/apache/arrow/pull/8662/checks?check_run_id=1399439929) ), so merging

**Comment by alamb** [comment]
FYI @rdettai , this PR has been merged
